fidesctl/src/fidesctl/core/generate_dataset.py

```diff
@@ -116,35 +134,143 @@ def create_dataset_collections(
+    are pulled from the server unless a manifest taxonomy is supplied.
     """
-    url = engine.url
-    name = url.database
-    dataset = Dataset(
-        fides_key=name,
-        name=name,
-        description=f"Fides Generated Description for Dataset: {name}",
-        collections=collections,
+    uncategorized_fields = []
+    total_field_count = 0
+    for db_dataset_key in db_collections.keys():
+        dataset = (
+            next(
+                (
+                    dataset
+                    for dataset in manifest_taxonomy.dataset
+                    if dataset.fides_key == db_dataset_key
+                ),
+                None,
+            )
+            if manifest_taxonomy
+            else get_server_resource(
+                url=url,
+                resource_type="dataset",
+                resource_key=db_dataset_key,
+                headers=headers,
+            )
+        )
+        db_dataset = db_collections.get(db_dataset_key, {})
+        (
+            current_uncategorized_keys,
+            current_field_count,
+        ) = find_uncategorized_dataset_fields(
+            dataset_key=db_dataset_key, dataset=dataset, db_dataset=db_dataset
+        )
+        total_field_count += current_field_count
+        uncategorized_fields += current_uncategorized_keys
+
+    return uncategorized_fields, total_field_count
+
+
+def database_coverage(
+    connection_string: str,
+    manifest_dir: Optional[str],
+    coverage_threshold: int,
+    url: AnyHttpUrl,
+    headers: Dict[str, str],
+) -> None:
+    """
+    Given a database connection string, fetches collections
+    and fields and compares them to existing datasets or datasets in a
+    local manifest (if one is provided).
+
+    Prints uncategorized fields and raises an exception if coverage
+    is lower than provided threshold.
+    """
+    manifest_taxonomy = parse(manifest_dir) if manifest_dir else None
+
+    # Generate the collections and fields for the target database
+    db_engine = get_db_engine(connection_string)
+    db_collections = get_db_collections_and_fields(db_engine)
+
+    uncategorized_fields, db_field_count = find_all_uncategorized_dataset_fields(
+        manifest_taxonomy=manifest_taxonomy,
+        db_collections=db_collections,
+        url=url,
+        headers=headers,
     )
-    return dataset
+    coverage_percent = (
+        int(((db_field_count - len(uncategorized_fields)) / db_field_count) * 100)
+        if db_field_count > 0
+        else 100
+    )
+
+    output: str = ""
+    if uncategorized_fields:
+        output += "The following fields do not have any data category annotations: \n"
+        output += "\n".join(uncategorized_fields) + "\n\n"
+    output += f"Datasets ({list(db_collections.keys())}) annotation coverage: {coverage_percent}%"
+
+    if coverage_percent < coverage_threshold:
+        echo_red(output)
+        raise SystemExit(1)
+    echo_green(output)
```

Comment:
```suggestion
    output = "Successfully scanned the following datasets:\n"
    output += f"\t{'\n\t'.join(list(db_collections.keys()).sort())}\n"
    print(output)
    
    if uncategorized_fields:
        uncategorized_output = "The following fields are missing data category annotations:\n"
        uncategorized_output += f"\t{'\n\t'.join(uncategorized_fields.sort())}\n"
        print(uncategorized_output)

    annotation_output = f"Annotation coverage: {coverage_percent}%\n\n"
    if coverage_percent < coverage_threshold:
        echo_red(annotation_output)
        raise SystemExit(1)
    echo_green(annotation_output)
```

This might need some tweaking. I'm going for output like:

```
Successfully scanned the following datasets:
    dataset_key_one
    dataset_key_three
    dataset_key_two

The following fields are missing data category annotations:
    uncategorized_field_one
    uncategorized_field_three
    uncategorized_field_two

Annotation coverage: 85%   # Either red or green, exits with a 1 status when red
```

Separately, and regardless of whether you feel this is an improved output format, it would be nice to extract the output construction into a helper function such that this consolidates to:

```python
print(
    results(db_collection.keys(), uncategorized_fields, coverage_percent, coverage_threshold)
)
```

Reply:
good change, updated